util/src/main/java/io/grpc/util/HealthProducerUtil.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright 2023 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.util;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static io.grpc.LoadBalancer.HAS_HEALTH_PRODUCER_LISTENER_KEY;
+import static io.grpc.LoadBalancer.HEALTH_CONSUMER_LISTENER_ARG_KEY;
+
+import io.grpc.Attributes;
+import io.grpc.ConnectivityStateInfo;
+import io.grpc.Internal;
+import io.grpc.LoadBalancer;
+
+/**
+ * Utility functions used by health producer systems to build health notification chain, via
+ * {@link io.grpc.LoadBalancer.CreateSubchannelArgs}.
+ * The leaf health consumer is pick first. Each health producer uses this helper.
+ * The health producers should make state listener a pass-through and manipulate the
+ * {@link io.grpc.LoadBalancer.CreateSubchannelArgs} for health notifications.
+ * The root health producer in the chain will fan out the subchannel state change to both
+ * state listener and health listener.
+ */
+@Internal
+public class HealthProducerUtil {
+
+  /**
+   * A new {@link io.grpc.LoadBalancer.Helper} that detects health listener parent and fans out the
+   * subchannel state change to both state listener and health listener.
+   *
+   * <p>Example usage:
+   * <pre>{@code
+   * class HealthProducerLB {
+   *   private final LoadBalancer.Helper helper;
+   *   public HealthProducer(Helper helper) {
+   *     this.helper = new MyHelper(HealthCheckUtil.HealthCheckHelper(helper));
+   *   }
+   *   class MyHelper implements LoadBalancer.Helper {
+   *     public void CreateSubchannel(CreateSubchannelArgs args) {
```

Comment:
This doesn't seem to be used.  Is non-standard to start the method name with a capital letter.